File: src/test/java/com/github/spartusch/rateretriever/rate/v1/service/WebQueryServiceImplTest.java
package com.github.spartusch.rateretriever.rate.v1.service;

import com.github.spartusch.webquery.WebQuery;
import com.github.spartusch.webquery.WebQueryFactory;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import java.nio.charset.StandardCharsets;

import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.entry;
import static org.mockito.Mockito.when;

public class WebQueryServiceImplTest {

    private WebQueryFactory factory;
    private WebQuery webQuery;

    private WebQueryServiceImpl service;

    @Before
    public void setUp() {
        webQuery = new WebQuery("content", StandardCharsets.UTF_8);
        factory = Mockito.mock(WebQueryFactory.class);
        service = new WebQueryServiceImpl(factory);
        when(factory.create("uri")).thenReturn(webQuery);
    }

    @Test
    public void test_getWebQueryEntity_content() {
        final var entity = service.getWebQueryEntity("uri", "symbol", "currencyCode");
        assertThat(entity.getBody()).isEqualTo(webQuery.getContent());
    }

    @Test
    public void test_getWebQueryEntity_contentType() {
        final var entity = service.getWebQueryEntity("uri", "symbol", "currencyCode");
        assertThat(entity.getHeaders()).contains(entry("Content-Type", singletonList(webQuery.getContentType())));
    }

    @Test
    public void test_getWebQueryEntity_contentLength() {
        final var entity = service.getWebQueryEntity("uri", "symbol", "currencyCode");
        assertThat(entity.getHeaders()).contains(entry("Content-Length", singletonList(String.valueOf(webQuery.getContentLength()))));
    }

    @Test
    public void test_getWebQueryEntity_contentDisposition() {
        final var entity = service.getWebQueryEntity("uri", "symbol", "currencyCode");
        final var expectedContentDisposition = webQuery.getContentDisposition("symbol_CURRENCYCODE.iqy");
        assertThat(entity.getHeaders()).contains(entry("Content-Disposition", singletonList(expectedContentDisposition)));
    }
}

File: src/test/kotlin/com/github/spartusch/rateretriever/rate/v1/service/WebQueryServiceImplTest.kt
package com.github.spartusch.rateretriever.rate.v1.service

import com.github.spartusch.rateretriever.rate.v1.model.TickerSymbol
import com.github.spartusch.webquery.WebQueryFactory
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.entry
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.core.io.ByteArrayResource
import javax.money.Monetary

class WebQueryServiceImplTest {

    private lateinit var cut: WebQueryServiceImpl

    private val symbol = TickerSymbol("symbol")
    private val currency = Monetary.getCurrency("EUR")

    @BeforeEach
    fun setUp() {
        cut = WebQueryServiceImpl()
    }

    @Test
    fun getWebQueryEntity_content() {
        val webQuery = WebQueryFactory.create("someUri")
        val entity = cut.getWebQueryEntity("someUri", symbol, currency)
        assertThat(entity.body).isEqualTo(ByteArrayResource(webQuery.contentBytes))
    }

    @Test
    fun getWebQueryEntity_contentType() {
        val entity = cut.getWebQueryEntity("someUri", symbol, currency)
        assertThat(entity.headers).contains(entry("Content-Type", listOf("text/plain; charset=UTF-8")))
    }

    @Test
    fun getWebQueryEntity_contentLength() {
        val webQuery = WebQueryFactory.create("someUri")
        val entity = cut.getWebQueryEntity("someUri", symbol, currency)
        assertThat(entity.headers).contains(entry("Content-Length", listOf(webQuery.contentLength.toString())))
    }

    @Test
    fun getWebQueryEntity_contentDisposition() {
        val entity = cut.getWebQueryEntity("someUri", symbol, currency)
        assertThat(entity.headers)
            .contains(entry("Content-Disposition", listOf("attachment; filename=symbol_EUR.iqy")))
    }
}
